Show error summary from API when transcript fails to load

robertknight opened the pull request:
The API returns JSON:API error objects with `code`, `title` and `details` fields. Show the `title` to the user instead of a generic "API call failed" message.

Part of https://github.com/hypothesis/via/issues/1028

**Testing:**

1. Go to http://localhost:9083/https://www.youtube.com/watch?v=e-43x-H4CNY
2. You should see the error message below:

<img width="913" alt="Transcript error display" src="https://github.com/hypothesis/via/assets/2458/90a6d71f-1a82-4a77-a445-d3522310b352">

Note that the error title and contents of the "Error details" box need a bunch of work on the backend to refine. For example subtitles are not so much "disabled" for this video as "not available" because the video doesn't have spoken words. This morning I also encountered the same error message for a video where the transcript just hadn't been generated. In the frontend we are just displaying what the backend sends.

robertknight committed:
Show error summary from API when transcript fails to load

The API returns JSON:API error objects with `code`, `title` and `details`
fields. Show the `title` to the user instead of a generic "API call failed"
message.

Part of https://github.com/hypothesis/via/issues/1028

robertknight commented on via/static/scripts/video_player/components/test/TranscriptError-test.js:
```diff
@@ -0,0 +1,33 @@
+    );
+    assert.equal(
+      wrapper.text(),
+      ['Unable to load transcript', 'Something went wrong'].join('')
```
`[...].join('')` is used here to separate out parts that are on different lines in the UI, but that don't have actual spaces between them.

robertknight commented on via/static/scripts/video_player/utils/api.ts:
```diff
@@ -57,7 +68,15 @@ export async function callAPI<T = unknown>(api: APIMethod): Promise<T> {
+      Array.isArray(resultData.errors) &&
+      resultData.errors.length > 0
+    ) {
+      error = resultData.errors[0];
```
The JSON:API format allows for multiple errors, but we currently only report the first one on the frontend. I'm not aware of a scenario where we'd expect the backend to report multiple yet.

robertknight commented on via/static/scripts/video_player/components/TranscriptError.tsx:
```diff
@@ -0,0 +1,20 @@
+          <p>{error.error.detail}</p>
+        </details>
+      )}
+    </div>
```
Styling is obvious not final here.

lyzadanger replied:
Yep, don't worry about that for now. I've got it.

lyzadanger commented:
@robertknight Can you help me understand the goal of the "details" box once it's more functional?

robertknight commented:
> @robertknight Can you help me understand the goal of the "details" box once it's more functional?

I was thinking it would work like the "Details" box in LMS error messages. It could include any technical information that we think might be useful for debugging or other purposes. In the LMS context, the support/success team for example will often open this if they are on a call with a user and inspect / make note of what is in there.

lyzadanger commented:
> I was thinking it would work like the "Details" box in LMS error messages. It could include any technical information that we think might be useful for debugging or other purposes. In the LMS context, the support/success team for example will often open this if they are on a call with a user and inspect / make note of what is in there.

Are you fairly convinced that it adds value? I know you want it to be tweaked, but right now it reads:

> Could not retrieve a transcript for the video https://www.youtube.com/watch?v=e-43x-H4CNY! This is most likely caused by: Subtitles are disabled for this video If you are sure that the described cause is not responsible for this error and that a transcript should be retrievable, please create an issue at https://github.com/jdepoix/youtube-transcript-api/issues. Please add which version of youtube_transcript_api you are using and provide the information needed to replicate the error. Also make sure that there are no open issues which already describe your problem!

AFAICT that doesn't add any information or use beyond what the concise error message says. Can you give me an example of what it might say in future that would be useful?

_Edited_: A plausible path forward could be to omit the `<details>` element until it has something useful to say. Looks like it's easy to drop in?

robertknight commented:
> AFAICT that doesn't add any information or use beyond what the concise error message says. Can you give me an example of what it might say in future that would be useful?

There is an open issue at https://github.com/hypothesis/via/issues/1044. What I am expecting as a result of resolving that issue is that the backend would omit the `detail` field for this particular error, and so the `<details>` field would not be shown.

We don't currently have a case where we have extra information to show. I would expect the backend to simply omit the `detail` field whenever it has nothing to add.

lyzadanger commented:
> There is an open issue at #1044. What I am expecting as a result of resolving that issue is that the backend would omit the `detail` field for this particular error, and so the `<details>` field would not be shown.

So, maybe we could punt on the `<details>` element for now until we have a clear need for it?

robertknight commented:
> So, maybe we could punt on the <details> element for now until we have a clear need for it?

OK, I'll do that.

robertknight committed:
Remove `<details>` element from errors until we have a use for it

We have decided to remove this for the moment as we don't have a good example
where it provides value. We'll re-add it when we do.

See https://github.com/hypothesis/via/pull/1030#issuecomment-1621730270

robertknight commented:
Done! - The `<details>` is now gone.

lyzadanger commented on via/static/scripts/video_player/components/test/TranscriptError-test.js:
```diff
@@ -0,0 +1,29 @@
+import TranscriptError from '../TranscriptError';
+
+describe('TranscriptError', () => {
+  it('displays just `Error.message` if there are no error details', () => {
```
Does this test description need adjustment in light of the `details` removal?

robertknight replied:
This is still valid as "error details" includes the title. This case is for when we don't have a server-generated error at all, or it doesn't include even a title.

lyzadanger commented on via/static/scripts/video_player/components/VideoPlayerApp.tsx:
```diff
@@ -459,9 +460,7 @@ export default function VideoPlayerApp({
-              <div data-testid="transcript-error">
-                Unable to load transcript: {transcript.message}
-              </div>
+              <TranscriptError error={transcript} />
```
I think it's neat that the same reference can accommodate either an actual transcript or an error (and am of the opinion that typing makes this possible versus a bug waiting to happen). But I admit I did a double-take at `error={transcript}` here.

robertknight replied:
Yeah, I might rename this to `transcriptOrError` in future.

lyzadanger commented on via/static/scripts/video_player/utils/test/api-test.js:
```diff
@@ -33,38 +33,71 @@ describe('callAPI', () => {
+    it('throws exception if result is not JSON', async () => {
+      fakeFetch
+        .withArgs(videoPlayerConfig.api.transcript.url)
+        .resolves(new Response('<b>Oh no</b>', { status: 500 }));
```
Just confirming that we intend to invent a 500 here in these cases :). Granted this is not new behavior in these changes, so feel free to ignore me.

robertknight replied:
Yes, this is mean to simulate something like eg. a 5xx "Bad Gateway" error if the server failed to respond within the various timeouts that the proxies in front have.

lyzadanger reviewed (approved):
Thanks for your patience! I'm working on collating some error-styling needs into a single ticket to track, so, I got that part.
